### PR TITLE
Multiple node creation

### PR DESCRIPTION
**Pull request by chrismguevara**
This adds the ability to create multiple nodes of the same configuration in one command. The hostnames of the new nodes will consist of the user-specified hostname joined with an increasing number (e.g. hostname-1, hostname-2, hostname-3). The hostnames are generated in a way that avoids naming collisions.

### CODE REVIEW

File: create/node.go [hunk 1]
```diff
@@ -230,3 +261,51 @@ func getBaseNodeTerraformConfig(terraformModulePath, selectedCluster string, sta
...
+	if nodesToAdd == 1 {
+		nodeNameUsed := false
+		for _, existingName := range existingNames {
+			if existingName == nodeName {
+				nodeNameUsed = true
+				break
+			}
+		}
+		if !nodeNameUsed {
+			return []string{nodeName}
```
Comment (nimajalali) [inline create/node.go]: Should we still append a `-NUMBER` here? What happens if they add another node later?

File: create/node.go [hunk 2]
```diff
@@ -230,3 +261,51 @@ func getBaseNodeTerraformConfig(terraformModulePath, selectedCluster string, sta
 
 	return cfg, nil
 }
+
+// Returns the hostnames that should be used when adding new nodes. Prevents naming collisions.
+func getNewHostnames(existingNames []string, nodeName string, nodesToAdd int) []string {
```
Comment (nimajalali) [inline create/node.go:273]: I know we don't have any unit tests for this project yet but this func might be the place to start. Thinking a table driven test to get through edge cases and such. Thoughts?

File: create/node.go [hunk 3]
```diff
@@ -207,7 +209,36 @@ func getBaseNodeTerraformConfig(terraformModulePath, selectedCluster string, sta
...
-	// TODO: Allow user to specify number of nodes to be created.
+	// Allow user to specify number of nodes to be created.
+	var countInput string
+	if viper.IsSet("node_count") {
+		countInput = viper.GetString("node_count")
+	} else {
+		prompt := promptui.Prompt{
+			Label: "Number of nodes to create",
+			Validate: func(input string) error {
+				_, err := strconv.ParseInt(input, 10, 64)
```
Comment (nimajalali) [inline create/node.go]: Should we verify it's a positive number or zero?

File: create/node.go [hunk 4]
```diff
@@ -207,7 +209,36 @@ func getBaseNodeTerraformConfig(terraformModulePath, selectedCluster string, sta
...
+		if err != nil {
+			return baseNodeTerraformConfig{}, err
+		}
+
+		countInput = result
+	}
+
+	// Verifying node count
+	nodeCount, err := strconv.Atoi(countInput)
+	if err != nil {
```
Comment (nimajalali) [inline create/node.go:241]: Same as above. Verify positive or zero

File: create/node_aws.go [hunk 5]
```diff
@@ -152,11 +152,28 @@ func newAWSNode(selectedClusterManager, selectedCluster string, remoteBackend ba
...
+	for nodeName := range nodes {
+		existingNames = append(existingNames, nodeName)
+	}
+
+	// Determine what the hostnames should be for the new node(s)
+	newHostnames := getNewHostnames(existingNames, cfg.Hostname, cfg.NodeCount)
+
+	// Add new node to terraform config with the new hostnames
+	for _, newHostname := range newHostnames {
+		cfgCopy := *(&cfg)
```
Comment (nimajalali) [inline create/node_aws.go]: Do we need this to create a copy of the struct?

I think `cfgCopy := cfg` would do the same. https://play.golang.org/p/XK_S0iQSnE

File: create/node_azure.go [hunk 6]
```diff
@@ -218,11 +218,28 @@ func newAzureNode(selectedClusterManager, selectedCluster string, remoteBackend
...
+	for nodeName := range nodes {
+		existingNames = append(existingNames, nodeName)
+	}
+
+	// Determine what the hostnames should be for the new node(s)
+	newHostnames := getNewHostnames(existingNames, cfg.Hostname, cfg.NodeCount)
+
+	// Add new node to terraform config with the new hostnames
+	for _, newHostname := range newHostnames {
+		cfgCopy := *(&cfg)
```
Comment (nimajalali) [inline create/node_azure.go]: Same copy cfg comment above

File: create/node_gcp.go [hunk 7]
```diff
@@ -216,11 +216,28 @@ func newGCPNode(selectedClusterManager, selectedCluster string, remoteBackend ba
...
+	for nodeName := range nodes {
+		existingNames = append(existingNames, nodeName)
+	}
+
+	// Determine what the hostnames should be for the new node(s)
+	newHostnames := getNewHostnames(existingNames, cfg.Hostname, cfg.NodeCount)
+
+	// Add new node to terraform config with the new hostnames
+	for _, newHostname := range newHostnames {
+		cfgCopy := *(&cfg)
```
Comment (nimajalali) [inline create/node_gcp.go]: Same copy cfg comment above

File: create/node_triton.go [hunk 8]
```diff
@@ -229,11 +229,28 @@ func newTritonNode(selectedClusterManager, selectedCluster string, remoteBackend
...
+	for nodeName := range nodes {
+		existingNames = append(existingNames, nodeName)
+	}
+
+	// Determine what the hostnames should be for the new node(s)
+	newHostnames := getNewHostnames(existingNames, cfg.Hostname, cfg.NodeCount)
+
+	// Add new node to terraform config with the new hostnames
+	for _, newHostname := range newHostnames {
+		cfgCopy := *(&cfg)
```
Comment (nimajalali) [inline create/node_triton.go]: Same copy cfg comment above

**Review by nimajalali — CHANGES_REQUESTED** [review]

File: create/node.go (hunk 3 above)
Comment (chrismguevara) [reply create/node.go]: Yeah, good point. I'll fix that.

File: create/node.go (hunk 2 above)
Comment (chrismguevara) [reply create/node.go:273]: Agreed. I'll add some tests to this.

File: create/node.go (hunk 1 above)
Comment (chrismguevara) [reply create/node.go]: I was trying to preserve the behavior where the user adds a single node and the hostname matches what the user specified. If they happen to add additional node(s) later those nodes would have the `-NUMBER` suffix resulting in a series of hostnames like `hostname, hostname-1, hostname-2, etc...`. Is it okay if we always add a `-NUMBER` for node hostnames even if they only want to add one node?

File: create/node_aws.go (hunk 5 above)
Comment (chrismguevara) [reply create/node_aws.go]: Oh interesting. I didn't know go makes a copy when you assign structs like that. I'll fix it.

Comment (nimajalali) [reply create/node_aws.go]: Yup, the only time this doesn't work is if the struct contains reference types. (pointers, slice, etc)

If the struct contained reference types go would just copy the reference not the full value.

File: create/node.go (hunk 1 above)
Comment (nimajalali) [reply create/node.go]: I think it's okay to always add a `-NUMBER`, even if there's only 1 we'll be consistent. Add it sets us up for good naming when the next node gets added.
